design/sticky-session-design.md

```diff
@@ -0,0 +1,170 @@
+# Sticky session support
+
+**Status**: _draft_
+
+This document proposes adding support for sticky session in contour.
+
+## Goals
+
+- Enabling support for sticky session in Contour using
+    - Cookie
+    - Source IP
+    - Header
+- Adding support in IngressRoute CRD. 
+
+## Non-goals
+
+- Making options in IngressRoute very specific to Envoy is not a goal. Options in IngressRoute will be generic.
+
+## Background
+
+Sticky-sessions allow requests to be routed to the same host based on certain parameters of the request.
+The parameters can be a cookie, the source IP, a custom header, etc.
+
+Envoy supports sticky sessions by using a hash-based load balancing strategy (RingHash, Maglev). 
+Specifying a *hash policy* for such load-balancers will enable sticky sessions. However, we are not aiming at making
+this very specific to Envoy. IngressRoute options should be generic enough without exposing
+internals of Envoy itself.
+
+
+## High-Level Design
+
+Sticky sessions are enabled per route. The following change are required in IngressRoute:
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata: 
+  name: basic
+  namespace: default
+spec: 
+  virtualhost:
+    fqdn: foo-basic.bar.com
+  routes: 
+    - match: /
+      sticky-session:
+        source: cookie/header/connection # There should be exactly one of these set.
+        name: name # This can be cookie name, header name, or connection property
+        ttl: time-to-live # This is only valid when source = cookie. 
+        path: cookie-path # This is only valid when source = cookie. 
+      services: 
+        - name: s1
+          port: 80  
+          strategy: RingHash # Optional. Default: RingHash
+```          
+
+For sticky sessions to be enabled on any route, the service associated to the route needs to have the 
+load balancing strategy set to a hash-based strategy (like RingHash or Maglev). The following cases need to be considered:
+
+- No service strategy is specified: In this case, contour will set the service's strategy to "RingHash" by default.
+- Service strategy specified, but not set to a Hash-based strategy: In this case, there will be no effect of declaring sticky session.
```

Comment:
I'd prefer that we didn't let users create valid -- ie, fit through validation -- configurations that don't do what they want. This puts more load on having to generate good status messages, which we're not doing very well in Contour at the moment, so that would be solving two problems at once. I'd prefer it if the CRD data structures tried to prevent users supplying invalid configuration.